### ISSUE TITLE
`while_used` checker shouldn't report `while True`

### ISSUE THREAD

**Issue by GideonBear** [PyCQA/pylint#7748]
### Current problem

Currently, the `while_used` checker reports `while-used` for a `while True` loop:
```py
from time import sleep

while True:
    sleep(1)
```
```
main.py:3:0: W0149: Used `while` loop (while-used)
```
The documentation says:
> Unbounded while loops can often be rewritten as bounded for loops.

Which is untrue for `while True` loops.
It is possible with an infinite generator, but `while True` makes much more sense.

### Desired solution

The `while_used` checker should not report an error for a `while True` loop.

### Additional context

`pylint --version` output:
```
pytpylint 2.15.5
astroid 2.12.12
Python 3.8.12
[GCC 10.3.0]
```

**Comment by Pierre-Sassoulas** [comment]
This goes to show that even the simplest of checker end up having some maintenance attached to it 😄

**Comment by GideonBear** [comment]
@Pierre-Sassoulas Can I try to fix this and submit a PR?

**Comment by Pierre-Sassoulas** [comment]
Sure thank you for proposing ! I'm assigning you to the issue :)

**Comment by jacobtylerwalls** [comment]
I'd like to invite more discussion on this. I thought of part of the point of `while-used` was to help opinionated code owners reject all uses of `while` no matter what. I'm not sure we should start carving out special cases.

**Comment by Pierre-Sassoulas** [comment]
I checked because I had the same afterthought than @jacobtylerwalls. Initial implementation was in #4367 and we did not talk about it BUT the same contributor implemented another opinionated check around the same time where we definitely had this argument. (mandatory ternary. We started to design complicated rules for when a ternary should be used exactly then concluded that we'd better keep it simple and opinionated: https://github.com/PyCQA/pylint/pull/4871#issuecomment-902250349). Maybe like in the ternary checker, it would be better to not touch anything and and keep it intentionally opinionated and simple for everyone sake (maintainers mostly). It's easy to disable the check if it's wrong on while True, the false negative on the contrary is impossible to fix on the user side.

On the other end the explanation for why exactly while loop are bad is not accurate in this case and ``while True`` is a fundamental design pattern of game loops, so I understand both side of the argument.

**Comment by jacobtylerwalls** [comment]
Exactly. My thought is that there are very few users of `while-used` that would want to carve out a huge exception for one of the most common use cases for `while`.

**Comment by GideonBear** [comment]
I agree that the plugin can be used to ban while loops entirely. The reason I thought of this as a false positive was because of the explanation of the error: "Unbounded while loops can often be rewritten as bounded for loops."
I don't know the internals of pylint, but would it be possible to either separate this into two error codes (so one can be ignored if wanted) or add a configuration option to allow constants?

**Comment by Pierre-Sassoulas** [comment]
The configuration option would be a middle-ground yes (default should be "don't allow constant", to not create a breaking change).

**Comment by orSolocate** [comment]
I agree with the configuration solution. It is more serious to have this code unintentionally than having this False Negative, when it rises as False Negative you just edit your config file and `pylint` skips it. We can explain in the docs that If you intend to have infinite loops you should add the `allow-const-while` flag in the config file, e.g. in gaming as mentioned or embedded applications.

**Comment by DanielNoord** [comment]
Can't you just do `for _ in itertools.count()`? If you're disallowing `while` loops (which imo are perfectly valid) you can expect to need to use some weird other patterns.

**Comment by GideonBear** [comment]
> Can't you just do `for _ in itertools.count()`? If you're disallowing `while` loops (which imo are perfectly valid) you can expect to need to use some weird other patterns.

That's true, but at least for me the point in banning while loops is to prevent using them where a for loop is more pythonic. IMO, `while True` is still more pythonic than `for _ in itertools.count()`. If you're incrementing a value each iteration, then yes, `itertools.count()` would be better, but that's not for every while loop.
That makes me think about it again, because making an exception for `while True` *would* allow incrementing. It's hard to make rules that work on everything...

**Comment by GideonBear** [comment]
> Yeah, which is why I don't think even a flag could make this as configurable as desired.

Then the simplest would be to just ban all while loops, so you can manually ignore the ones you know can't be rewritten as for loops, but then we're right back where we started.

> If you're really looking for a specific type of check I think it might be better to create your own extension, which you can base on the general while-checker.

Maybe there could be a configuration flag (something like `while-used-strict`) which is on by default (or `while-used-loose` that's off by default, I don't know), that if on reports every use of `while`, and if it's off, it reports only usages of while it thinks can be rewritten as for loops (non-constant test, any in-place operator on a variable (`+=`, `*=`, `-=` etc) in the body)? That way it can still be used as a full ban, but can also be used to check if it may be possible to use a for loop, but doesn't ban them entirely.

This sounds like it fits in pylint (as it already has a `while-used` extension), but if this isn't accepted (which I would understand) I'm going to make my own `flake8` plugin for it (as I use `flake8-pylint`)

**Comment by DanielNoord** [comment]
> Then the simplest would be to just ban all while loops, so you can manually ignore the ones you know can't be rewritten as for loops, but then we're right back where we started.

Imo, that makes the most sense actually. If you consciously disallow legitimate features of Python I think it is to be expected that you need to add some disables somewhere.

> This sounds like it fits in pylint (as it already has a `while-used` extension), but if this isn't accepted (which I would understand) I'm going to make my own `flake8` plugin for it (as I use `flake8-pylint`)

I'm starting to wonder if we should have this extension in the first place. We accepted it as we thought it wouldn't create much maintenance cost as it was relatively simple. This issue shows that that might not actually be the case. I'd be in favour of keeping this as simple as possible and let any more advanced `while` checking be part of a separate extension.
If that extension ever gets created we could even consider linking to it and removing the base extension ourselves.

**Comment by Pierre-Sassoulas** [comment]
I think the checker won't see much more change anymore if we add that option, and we already invested some time in discussing this. Let's put it to a vote:

- While checker simple and opinionated 🚀 
- Add an option to permit to allow while with constants 👀

**Comment by nickdrozd** [comment]
Hello! I proposed and implemented the extension, so maybe I can weigh in.

I am sympathetic to adding an exception for `while True`. Really it would be best if there was something like a `loop` keyword that just did that. `while True` is just a crummy hack. (It could be worse though; some really old code in the CPython code still uses `while 1`.) If we had a `loop` operator, then we would all be in agreement that it should be used in place of `while True`, and there would be a check for that.

Still, this is an opt-in extension. Nobody is being forced to play in strict mode here. My feeling is that anyone who is sophisticated enough to enable this check is also sophisticated enough to add `pylint: disable = while-used` when there really is a good reason to use it. There are not many occasions where this is actually the case, so this shouldn't be much of an intrusion.

For that same reason, the cases where a `while` really does have to be used tend to be very interesting places in the code, so grepping for `while-used` can be a great way to navigate a codebase!

Now, I don't work on games, and maybe `while` loops are used all over the place there. Game developers probably shouldn't enable this check! But in my experience most `while` loops can be expressed nicer in other ways, or at least wrapped in a generator as in https://github.com/PyCQA/astroid/pull/1169.

So: I would like the check to be kept around, because I, as a user of Pylint, really like it. I wish Python had a `loop` keyword to express intentionally infinite loops, but given that it doesn't, I think it's reasonable to add an exception for `while True`. (What about `while 1` or `while 'string'`?)

I will abstain from the vote :cat:

**Comment by nickdrozd** [comment]
As an alternative, I fully support anyone who wants to change to documentation to list exceptions.

For example: "Unbounded while loops can often be rewritten as bounded for loops. Exceptions include event loops, listeners, ... [etc]."

**Comment by DanielNoord** [comment]
Seeing the result of the poll, I think we can close this? We might want to update the documentation, but it seems consensus is to keep the checker simple so we lower the maintenance burden?

**Comment by Pierre-Sassoulas** [comment]
Let simply upgrade the documentation / message description before closing this one.

**Comment by Pierre-Sassoulas** [comment]
@GideonBear I unnassigned you because what need to be done changed significantly but you're welcome to upgrade the documentation if you want :)
